pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -907,44 +910,50 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 			if hashedServices[firstHash] == nil {
 				hashedServices[firstHash] = svc
 			} else {
-				// This means we have a collision. Resolve collision by "DoubleHashing".
-				i := uint32(1)
-				secondHash := uint32(prime) - (s % uint32(prime))
-				for {
-					nh := (s + i*secondHash) % uint32(maxIPs-1)
-					if hashedServices[nh] == nil {
-						hashedServices[nh] = svc
-						break
-					}
-					i++
-				}
+				hashCollisions = append(hashCollisions, svc)
 			}
 			hash.Reset()
 			j++
 		}
 	}
 
-	x := 0
-	hnMap := make(map[string]octetPair)
-	for _, svc := range hashedServices {
+	// rehash at last, so that it can mitigate the influcence to other services that donot have hash collision
+	// It maybe not good the these services, but it is beneficial to most other services without hash collision.
+	for _, svc := range hashCollisions {
+		// First hash is calculated by hashing the service key i.e. (namespace + "/" + hostname).
+		hash.Write([]byte(makeServiceKey(svc)))
+		s := hash.Sum32()
+		hash.Reset()
+		// This means we have a collision. Resolve collision by "DoubleHashing".
+		i := uint32(1)
+		secondHash := uint32(prime) - (s % uint32(prime))
+		for {
+			nh := (s + i*secondHash) % uint32(maxIPs-1)
+			if hashedServices[nh] == nil {
+				hashedServices[nh] = svc
+				break
+			}
+			i++
+		}
+	}
+
+	for i, svc := range hashedServices {
 		if svc == nil {
-			// There is no service in the slot. Just increment x and move forward.
-			x++
+			// There is no service in the slot.
 			continue
 		}
+		if svc.AutoAllocatedIPv4Address != "" {
```

Comment:
Again this might cause the split brain issue?

Reply:
Sorry i forget to remove it